source/MicroBitMemoryMap.cpp

```diff
@@ -138,5 +142,64 @@ void MicroBitMemoryMap::findHashes()
                 memoryMapStore.memoryMap[2].startAddress = (uint32_t)magicAddress;
                 return;
         }
+        
+        if (   *(magicAddress - (sizeof(magicAddress))) == 0x597F30FE
+            && *(magicAddress - (sizeof(magicAddress)/4))   == 0xC1B1D79D
+           ) {
+            // Found uPy Magic
+            DMESG("FOUND UPY_MAGIC");
+            uint8_t nRegions = (*(magicAddress - (sizeof(magicAddress)/2)));
+            
+            // Iterate through regions
+            for(int i = 0; i < nRegions; i++) {
+               processRecord(magicAddress - ((2 + i) * sizeof(magicAddress))); 
+            }
+
+        }
+    }
+}
+
+
+/*
+ * Function to process record from uPy build
+ */
+void MicroBitMemoryMap::processRecord(uint32_t *address) {
+
+    MicroPythonLayoutRecord record;
+    memcpy(&record, address, sizeof(record));
+
+    uint32_t start = record.reg_page * MICROBIT_CODEPAGESIZE;
+
+    uint32_t length = record.reg_len;
+
+    DMESG("Python Layout Record. Record: %d Hash Type: %d Start: %x Length: %x Hash: %x %x"
+                    , record.id
+                    , record.ht
+                    , start
+                    , length
+                    , record.hash_data[1]
+                    , record.hash_data[0]);
+
+    uint32_t hash[2]; 
+    if(record.ht == HASH_TYPE_POINTER) {
+        // Points to a string
+        // Take string and generate 8 byte hash
+        uint8_t * string = (uint8_t *) record.hash_data[0];
+        int n   = 0;
+        while(n < MAX_STRING_LENGTH && *(string + n) != '\0') {
+            n++;
+        }
+
+        uint32_t crc = crc32_compute(string, n, NULL);
```

Comment:
Is a 4 byte CRC good enough for this?

Reply:
The string is :
`micro:bit v2.0.0-beta.1+130a254-dirty on 2020-12-15; MicroPython v1.13 on 2020-12-15`

So we'd have to pick the 4 byte git hash from the string, (or could event extract it and use it?)

Reply:
Oh, do you mean the resulting CRC from the operation?
I think we have space for 8 byte hashes, so it could halve the string and generate a hash for each half

Reply:
I'm not saying the CRC isn't good enough. A false match probably wouldn't brick micro:bit. I think it would be more sensible if the python tool that creates the table made eight bytes that identify the micropython-microbit-v2 build. Maybe the UTC of the build would be OK for most uses, though inconvenient for testing.

The string is
```
"micro:bit v" MICROBIT_RELEASE "+" MICROBIT_GIT_HASH " on " MICROBIT_BUILD_DATE \
    "; MicroPython " MICROPY_GIT_TAG " on " MICROPY_BUILD_DATE
``` 

I'm not sure where MICROBIT_GIT_HASH and MICROPY_GIT_TAG come from, or what format we could rely on.

Reply:
So along the lines of?
It will defo avoid the collision issue, but you're right it will be annoying for testing!
```
--- a/src/addlayouttable.py
+++ b/src/addlayouttable.py
@@ -46,6 +46,7 @@ import argparse
 import binascii
 import struct
 import sys
+import time

 IHEX_TYPE_DATA = 0
 IHEX_TYPE_EXT_LIN_ADDR = 4
@@ -209,7 +210,7 @@ def main():
     layout = FlashLayout()
     layout.add_region(1, sd_start, sd_end - sd_start, FlashLayout.REGION_HASH_NONE)
     layout.add_region(
-        2, mp_start, mp_end - mp_start, FlashLayout.REGION_HASH_PTR, mp_version
+        2, mp_start, mp_end - mp_start, FlashLayout.REGION_HASH_DATA, hex(int(time.time()))
     )
     layout.add_region(3, fs_start, fs_end - fs_start, FlashLayout.REGION_HASH_NONE)
     layout.finalise()
```

Reply:
It was just an idea. I don't know what hex(int(time.time()) produces, or exactly what time.time() does for that matter! As released versions wouldn't be generated too rapidly, maybe one second precision is not required.

MICROBIT_BUILD_DATE and MICROPY_BUILD_DATE seem irrelevant as to whether a full flash is required.

I suppose MICROBIT_GIT_HASH and MICROPY_GIT_TAG aren't available to python? And I can't see how to make 8 bytes from them.

I don't think we've ever had a problem with the hash MakeCode uses, but it ought to be possible to make a simple version number for the python hex.

Reply:
I spoke to Carlos and Jonny and they would prefer to keep the pointer, it provides easy access to the version string for any program that can parse the hex

So if we keep the pointer and version string, should we make this 8 bytes by splitting it in two? Or using a 64bit hash? Or keep it simple and as it is with the CRC32

Reply:
Maybe keep it simple and see if we notice a problem?

I would change
```
        uint32_t crc = crc32_compute(string, n, NULL);
        memcpy(&hash, &crc, 4);

```
to 
```
        hash[0] = crc32_compute(string, n, NULL);
        hash[1] = 0;

```

Reply:
How do you calculate the CRC in the Android app?